### DISCUSSION TITLE
AutoARIMA update behaviour

### DISCUSSION THREAD

**Discussion by h-tw** [alan-turing-institute/sktime#3046] [Q&A]
Hi all.

I am currently playing around with the `AutoARIMA` forecaster and I am somewhat irritated that calling `update` just defaults to refitting the forecaster. 

I use a temporal train-test split on my data to run a model tournament between multiple estimators. One of the estimators is the `AutoARIMA` forecaster.  If I understand correctly `AutoARIMA` is in essence akin to an advanced grid search for the best ARIMA model over a defined parameter space. In other words, the result I get after fitting the forecaster is an ARIMA model of a specific order (say: ARIMA(0,1,1)) with specific coefficient values. 

Assume my model tournament ends with the `AutoARIMA` forecaster being chosen as the best overall estimator. What I would want to do next is update the best forecaster with the full data (including the test set). But updating from my initial perspective would mean updating the model coefficients without touching the ARIMA order - but this is not how things go at the moment. Instead of just updating the coefficients the `AutoARIMA` search for the best model is started anew, essentially refitting the forecaster. When staying within the grid search analogy this in essence means retuning the hyperparameters (AR, I and MA order). This to me feels odd.

My questions:
- Am I missing something important? Is a complete refit actually what I want in the outlined case/what you always want? Why would that be?
- If I just want to update the coefficients, how would I go about it? Fitting an `AutoARIMA` forecaster first, extracting the model order and than fit a regular `ARIMA` model? Is there an easier way?

Disclaimer: I am aware that sktime is just wrapping the pmdarima implementation. However, I wanted to raise this here foremost as conceptual issue.

Best htw

**Comment by fkiraly** [comment]
Thanks @h-t-w for raising this!

There are a number of issues linked.

> Am I missing something important? Is a complete refit actually what I want in the outlined case/what you always want? Why would that be?

1. This is the default, if not a different specific `_update` is implemented. I think it is the most sensible default when the developer does not implement a specific `_update`.

2. Whether `AutoARIMA` should do sth else in `update`, and if yes, what, is not so clear to me.

From an interface perspective, `update` of a given estimator imo it should update all fittable parameters, and that includes the pdq, since in `AutoARIMA` they are actually fitted.

A related issue on "ARIMA should have a specific `update`" is here: 
https://github.com/alan-turing-institute/sktime/issues/1788

3. Now, the number 2 default (all "fittable parameters" are updated) may not actually what you want, because you want to update only the coefficients, not the pdq. Do I understand correctly?

There might not be a current way to do this - there are some compositors in the `forecasting.stream` modules that let you allow how often you update and how often you refit, but this does not apply to different components separately, so you could not apply this, say, not to the grid search, but to the ARIMA inside, even if you build your modular AutoARIMA yourself.

We could build sth that allows you to update the outer parameters pdq less often (e.g., never), say, via this parameter estimator module:
https://github.com/alan-turing-institute/sktime/pull/3041

That would require building a `ParamFitter` estimator that estimates pdq and does nothing else. Then we chain that to ARIMA, and turn off updating in the pdq extractor.
I could build the first component easily if you know of a piece of code that produces pdq without making a forecast etc. Does this exist somewhere in `pmdarima`?

**Reply by h-tw** [reply]
> 1. This is the default, if not a different specific `_update` is implemented. I think it is the most sensible default when the developer does not implement a specific `_update`.

Totally agree.

> 2. Whether `AutoARIMA` should do sth else in `update`, and if yes, what, is not so clear to me.
> 
> From an interface perspective, `update` of a given estimator imo it should update all fittable parameters, and that includes the pdq, since in `AutoARIMA` they are actually fitted.
> 

I can definitely see this argument - just having the option to update coefficients as suggested above would be great. Conceptually, forecasters like `AutoARIMA`, `AutoETS`, `TBATS` are imo somewhat special because they not only solve for the best coefficients, but meta-optimize for number/relation of model parameters. Nonetheless, for maintaining a coherent interface you probably want the current update behavior as default.

> 3. Now, the number 2 default (all "fittable parameters" are updated) may not actually what you want, because you want to update only the coefficients, not the pdq. Do I understand correctly?

Correct, I would like to have this an option. 

> We could build sth that allows you to update the outer parameters pdq less often (e.g., never), say, via this parameter estimator module: #3041
> 
> That would require building a `ParamFitter` estimator that estimates pdq and does nothing else. Then we chain that to ARIMA, and turn off updating in the pdq extractor. I could build the first component easily if you know of a piece of code that produces pdq without making a forecast etc. Does this exist somewhere in `pmdarima`?

Sorry, but I am not really following here. What would I/we gain by this approach in comparison to?:

> Alternative could be this: #3051
> 
> That would allow you to build sth similar to what you want as `ForecastingGridSearch(ARIMA(), cv, pdq_grid, update_best_params=False)`

**Reply by fkiraly** [reply]
> What would I/we gain by this approach in comparison to

Approach 1 splits the "Auto" part off ARIMA, and allows to combine it with anything that needs p,d,q, e.g., the different ARIMA implementations. So it is more flexible in having different options to find p,d,q.

Approach 2 applies only to grid search, but allows to separate update behaviour of the outer from update behaviour of the inner.

The two approaches cover somewhat different general cases, but both allow to cover the specific case (separating update behaviour of pdq from that of ARIMA coeffs).

**Comment by fkiraly** [comment]
Alternative could be this:
https://github.com/alan-turing-institute/sktime/issues/3051

That would allow you to build sth similar to what you want as
`ForecastingGridSearch(ARIMA(), cv, pdq_grid, update_best_params=False)`

**Reply by h-tw** [reply]
> Alternative could be this: #3051
> 
> That would allow you to build sth similar to what you want as `ForecastingGridSearch(ARIMA(), cv, pdq_grid, update_best_params=False)`

@fkiraly 
Thanks for all the useful input.

At first I wasn't sure if this would solve my specific problem, for somewhat related reasons. Given the length of my time series, I prefer to opt for a simple train-test-split instead of cross validation. As far as I can tell `ForecastingGridSearchCV` is requiring a `cv` parameter. However, it seams one can pass as `SingleWindowSplitter` with just one window (`window_length = len(time_series) - fh`) to `ForecastingGridSearchCV`, de facto circumventing the cross validation component. This is good.

What your suggesting would actually be somewhat akin to calling `AutoARIMA` with `stepwise=False`, right? With the benefit - when implementing your suggestion - that one could update the coefficients, without changing `p, d, q`. This would of course be slower, than `AutoARIMA` with defaults, but I guess that would be an acceptable trade-off for my current use case. And one would loose some other benefits (e.g. automated differencing...). 

In sum, I personally think this would be a great enhancement in general, I am however not totally convinced that the best approach - at least for my scenario - is working around `AutoARIMA`.

**Reply by fkiraly** [reply]
If you do not want to work around `AutoARIMA`, the downstream method in `pmdarima` would need to offer that precise option, i.e., not updating pdq but updating coefficients. Does that exist? If yes, surely it must be accessible through some of the parameter combinations exposed in `sktime` since it aims to expose all parameters. (or this would be sth to add since it was mistakenly left out)

**Comment by fkiraly** [comment]
>However, I wanted to raise this here foremost as conceptual issue.

Yes, nice one! Any opinions on the above?

**Comment by h-tw** [comment]
I was just cruising around the pmdarima documentation and code.

Their `AutoARIMA` implementation inherits from the `ARIMA` class and this class has apparently an `update` method.  And this method does only update the coefficients, without touching the p,d,q values - doing exactly what I would like as an option.

For an example see [update example](https://alkaline-ml.com/pmdarima/auto_examples/arima/example_add_new_samples.html#adding-new-observations-to-your-model).

This should make things easier, or not? For starters, I guess the class `_PmdArimaAdapter` would need a new, additional `update` method?

**Reply by fkiraly** [reply]
yes, an `_update` method, to be precise.

Imo we still should provide the current default behaviour, where pdq are also updated, so I would recommend to add a hyperparameter `update_pdq: boolean`, and do the one or the other, depending what the hyperparameter says.

**Reply by h-tw** [reply]
Among all the options we have discussed so far, this would be the solution I would go for. And agreed to providing current default.

If you want, I can give it a shot. I don't have so much experience in contributing code to open source, but this seams a rather straight forward case.

**Reply by fkiraly** [reply]
sorry, didn't see this! Would have appreciated the contribution - apologies for grabbing it instead of giving you the opportunity to try.

Either way, `update` is missing in a few other places too, so that's also a good place to start!

**Reply by fkiraly** [reply]
like this? https://github.com/alan-turing-institute/sktime/issues/3065

**Reply by h-tw** [reply]
No worry, all good.
